### PR TITLE
Ac/quantity

### PR DESCRIPTION
**Pull request by croysdaa**
This is the items list with quantities:
image
Edit Item page with quantity field including current value for quantity:
image

### CODE REVIEW

File: back-end/src/main/java/app/cluttermap/service/ItemsService.java [hunk 1]
```diff
@@ -50,6 +50,9 @@ public Item updateItem(Long id, UpdateItemDTO itemDTO) {
         if (itemDTO.getDescription() != null) {
             _item.setDescription(itemDTO.getDescription());
         }
+        if (itemDTO.getQuantity() != null) {
```
Comment (croysdar) [inline back-end/src/main/java/app/cluttermap/service/ItemsService.java:53]: You should add something to make sure the quantity is more than 0

If it is less than 0 you need to throw an exception

File: back-end/src/test/java/app/cluttermap/controller/ItemControllerTests.java [hunk 2]
```diff
@@ -247,8 +250,8 @@ void addOneItem_ShouldReturnBadRequest_WhenOrgUnitIdIsNaN() throws Exception {
     void updateOneItem_ShouldUpdateItem_WhenValidRequest() throws Exception {
         // Arrange: Set up an UpdateItemDTO with a new name and mock the service to
         // return the updated item
-        UpdateItemDTO itemDTO = new UpdateItemDTO("Updated Item", "Updated Description", List.of("Updated Tag"));
-        Item updatedItem = new Item(itemDTO.getName(), itemDTO.getDescription(), itemDTO.getTags(), mockOrgUnit);
+        UpdateItemDTO itemDTO = new UpdateItemDTO("Updated Item", "Updated Description", List.of("Updated Tag"), 2);
+        Item updatedItem = new Item(itemDTO.getName(), itemDTO.getDescription(), itemDTO.getTags(), mockOrgUnit, 1);
```
Comment (croysdar) [inline back-end/src/test/java/app/cluttermap/controller/ItemControllerTests.java:254]: instead of 1 this should be itemDTO.getQuantity()

**Comment by croysdar** [comment]
Make sure you add a tests in the ItemControllerTest to test if you give it a string value or a null value for the quantity that it returns an exception

**Comment by croysdar** [comment]
I would also add a second constructor that doesn't contain quantity but sets quantity to 1 as a default. You don't necessarily have to have all the changes in the tests in that case.

File: back-end/src/main/java/app/cluttermap/model/Item.java [hunk 3]
```diff
@@ -34,17 +34,20 @@ public class Item {
...
     // public constructor
     // ID is not required because Postgres generates the ID
-    public Item(String name, String description, List<String> tags, OrgUnit orgUnit) {
+    public Item(String name, String description, List<String> tags, OrgUnit orgUnit, Integer quantity) {
         this.name = name;
         this.description = description;
         this.orgUnit = orgUnit;
         this.tags = tags;
+        this.quantity = quantity;
     }
```
Comment (croysdar) [inline back-end/src/main/java/app/cluttermap/model/Item.java:52]: Try adding a second constructor

public Item(String name, String description, List<String> tags, OrgUnit orgUnit) {
        this.name = name;
        this.description = description;
        this.orgUnit = orgUnit;
        this.tags = tags;
        this.quantity = 1;
}

File: back-end/src/test/java/app/cluttermap/service/ItemServiceTests.java [hunk 4]
```diff
@@ -104,11 +104,11 @@ void createItem_ShouldCreateItem_WhenRoomExists() {
         when(orgUnitService.getOrgUnitById(1L)).thenReturn(mockOrgUnit);
 
         // Arrange: Prepare the Item DTO with the room ID as a string
-        NewItemDTO itemDTO = new NewItemDTO("New Item", "Item description", List.of(), String.valueOf(1L));
+        NewItemDTO itemDTO = new NewItemDTO("New Item", "Item description", List.of(), String.valueOf(1L), 1);
 
         // Arrange: Create a mock Item that represents the saved item returned by
         // the repository
-        Item mockItem = new Item(itemDTO.getName(), itemDTO.getDescription(), List.of(), mockOrgUnit);
+        Item mockItem = new Item(itemDTO.getName(), itemDTO.getDescription(), List.of(), mockOrgUnit, 1);
```
Comment (croysdar) [inline back-end/src/test/java/app/cluttermap/service/ItemServiceTests.java:111]: change the 1 to itemDTO.getQuantity()

File: react-front-end/src/features/items/AddItem.tsx [hunk 5]
```diff
@@ -90,6 +92,18 @@ export const AddItem = () => {
                         InputLabelProps={{ shrink: true }}
                     />
 
+                    {/* Item Quantity */}
```
Comment (croysdar) [inline react-front-end/src/features/items/AddItem.tsx:95]: Should we have a default value of one here? Or is it fine since it is not required (as long as we make those other changes to the constructor I suggested)

File: react-front-end/src/features/items/EditItem.tsx [hunk 6]
```diff
@@ -104,6 +106,22 @@ const EditItem = () => {
                         variant="outlined"
                         InputLabelProps={{ shrink: true }}
                     />
+                    <Typography gutterBottom align="left">
+                        {/* Item Quantity */}
+                        <TextField
```
Comment (croysdar) [inline react-front-end/src/features/items/EditItem.tsx:111]: Might want to make it required on this one so that people don't wipe out the item quantity when editing

File: react-front-end/src/features/items/ItemsAccordion.tsx [hunk 7]
```diff
@@ -78,7 +78,10 @@ const ItemsAccordion: React.FC<ItemsAccordionProps> = ({ orgUnitId }) => {
                                     sx={{py: 1}}
                                 />
                                 <CardContent sx={{py: 1}}>
-                                    <Typography variant="body2">{item.description}</Typography>
+                                    <Typography variant="body2" gutterBottom align="left">{item.description}</Typography>
```
Comment (croysdar) [inline react-front-end/src/features/items/ItemsAccordion.tsx:81]: Both typographys should probably be in the same CardContent. A Card shouldn't really have multiple CardContents

**Comment by croysdaa** [comment]
> Make sure you add a tests in the ItemControllerTest to test if you give it a string value for the quantity that it returns an exception

I'm not sure how to approach this issue. The constructor only accepts integers for the quantity value

File: back-end/src/main/java/app/cluttermap/model/dto/NewItemDTO.java [hunk 8]
```diff
@@ -19,11 +19,22 @@ public class NewItemDTO {
     @Pattern(regexp = "\\d+", message = "OrgUnit ID must be a valid number.")
     private String orgUnitId;
 
+    private Integer quantity;
```
Comment (croysdar) [inline back-end/src/main/java/app/cluttermap/model/dto/NewItemDTO.java:22]: Add these


    @NotNull(message = "The value cannot be null")
    @Min(value = 0, message = "The value must be at least 0")
    
    And then you can make tests to make sure those messages pop up in cases where incorrect input is given.
    
    Similar to this test:
    

    @Test
    void updateOneOrgUnit_ShouldReturnBadRequest_WhenOrgUnitNameIsNull() throws Exception {
        // Arrange: Set up an UpdateOrgUnitDTO with a null orgUnit name to trigger
        // validation
        UpdateOrgUnitDTO orgUnitDTO = new UpdateOrgUnitDTO(null, "Updated Description");

        // Act: Perform a PUT request to the /org-units/1 endpoint with the null orgUnit
        // name
        mockMvc.perform(put("/org-units/1")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(orgUnitDTO)))
                .andExpect(status().isBadRequest())

                // Assert: Verify the validation error response for the name field
                .andExpect(jsonPath("$.errors[0].field").value("name"))
                .andExpect(jsonPath("$.errors[0].message").value("Organization unit name must not be blank."));
    }

File: back-end/src/test/java/app/cluttermap/service/ItemServiceTests.java [hunk 9]
```diff
@@ -290,6 +292,7 @@ void updateItem_ShouldNotChangeTags_WhenTagsIsNull() {
         assertThat(updatedItem.getName()).isEqualTo("Updated Name");
         assertThat(updatedItem.getDescription()).isEqualTo("Updated Description");
         assertThat(updatedItem.getTags()).isEqualTo(List.of("tag 1", "tag 2"));
+        assertThat(updatedItem.getQuantity()).isEqualTo(1);
```
Comment (croysdar) [inline back-end/src/test/java/app/cluttermap/service/ItemServiceTests.java:295]: make sure this matches the previous quantity (2)

File: back-end/src/test/java/app/cluttermap/service/ItemServiceTests.java [hunk 10]
```diff
@@ -268,20 +269,21 @@ void updateItem_ShouldNotChangeDescription_WhenDescriptionIsNull() {
         assertThat(updatedItem.getName()).isEqualTo("Updated Name");
         assertThat(updatedItem.getDescription()).isEqualTo("Initial Description");
         assertThat(updatedItem.getTags()).isEqualTo(List.of("Updated tag 1", "Updated tag 2"));
+        assertThat(updatedItem.getQuantity()).isEqualTo(1);
```
Comment (croysdar) [inline back-end/src/test/java/app/cluttermap/service/ItemServiceTests.java:272]: Make sure this matches the updated quantity (2)

**Review by croysdar — DISMISSED** [review]
